comps/dataprep/redis/multimodal_langchain/prepare_videodoc_redis.py

```diff
@@ -0,0 +1,533 @@
+        lb_ingesting = max(0, i - num_transcript_concat_for_ingesting)
+        ub_ingesting = min(len(annotation), i + num_transcript_concat_for_ingesting + 1)
+        caption_for_ingesting = " ".join([annotation[j]["caption"] for j in range(lb_ingesting, ub_ingesting)])
+
+        # augment this frame's transcript with more neighboring frames' transcript to provide more context to LVM for question answering
+        lb_inference = max(0, i - num_transcript_concat_for_inference)
+        ub_inference = min(len(annotation), i + num_transcript_concat_for_inference + 1)
+        caption_for_inference = " ".join([annotation[j]["caption"] for j in range(lb_inference, ub_inference)])
+
+        video_id = frame["video_id"]
+        b64_img_str = frame["b64_img_str"]
+        time_of_frame = frame["time"]
+        embedding_type = "pair"
+        source_video = frame["video_name"]
+
+        text_list.append(caption_for_ingesting)
+        image_list.append(path_to_frame)
+        metadatas.append(
+            {
+                "content": caption_for_ingesting,
+                "b64_img_str": b64_img_str,
+                "video_id": video_id,
+                "source_video": source_video,
+                "time_of_frame_ms": float(time_of_frame),
+                "embedding_type": embedding_type,
+                "title": title,
+                "transcript_for_inference": caption_for_inference,
+            }
+        )
+
+    return text_list, image_list, metadatas
+
+
+def ingest_multimodal(videoname, data_folder, embeddings):
+    """Ingest text image pairs to Redis from the data/ directory that consists of frames and annotations."""
+    data_folder = os.path.abspath(data_folder)
+    annotation_file_path = os.path.join(data_folder, "annotations.json")
+    path_to_frames = os.path.join(data_folder, "frames")
+
+    annotation = load_json_file(annotation_file_path)
+
+    # prepare data to ingest
+    text_list, image_list, metadatas = prepare_data_and_metadata_from_annotation(annotation, path_to_frames, videoname)
+
+    MultimodalRedis.from_text_image_pairs_return_keys(
+        texts=[f"From {videoname}. " + text for text in text_list],
+        images=image_list,
+        embedding=embeddings,
+        metadatas=metadatas,
+        index_name=INDEX_NAME,
+        index_schema=INDEX_SCHEMA,
+        redis_url=REDIS_URL,
+    )
+
+
+def drop_index(index_name, redis_url=REDIS_URL):
+    print(f"dropping index {index_name}")
+    try:
+        assert Redis.drop_index(index_name=index_name, delete_documents=True, redis_url=redis_url)
+        print(f"index {index_name} deleted")
+    except Exception as e:
+        print(f"index {index_name} delete failed: {e}")
+        return False
+    return True
+
+
+@register_microservice(
+    name="opea_service@prepare_videodoc_redis", endpoint="/v1/dataprep/generate_transcripts", host="0.0.0.0", port=6007
+)
+@traceable(run_type="tool")
```

Comment:
please remove langsmith

Reply:
Thank for your feedback. We have removed langsmith here.